File: app/src/main/java/com/example/cobe/cryptonews/ui/articleSearch/ArticlesSearchActivity.java
package com.example.cobe.cryptonews.ui.articleSearch;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.cobe.cryptonews.interaction.ArticlesInteractorImpl;
import com.example.cobe.cryptonews.interaction.ArticlesInteractorInterface;
import com.example.cobe.cryptonews.presentation.SearchContract;
import com.example.cobe.cryptonews.presentation.implementation.SearchPresenter;
import com.example.cobe.cryptonews.ui.articles.ArticleAdapter;
import com.example.cobe.cryptonews.R;
import com.example.cobe.cryptonews.api.ApiClient;
import com.example.cobe.cryptonews.listeners.OnArticleClickListener;
import com.example.cobe.cryptonews.model.Article;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ArticlesSearchActivity extends AppCompatActivity implements OnArticleClickListener, SearchContract.View {

    @BindView(R.id.rvArticlesSearchList)
    RecyclerView recyclerView;

    @BindView(R.id.tvSearchWord)
    TextView searchedWord;

    @BindView(R.id.tvSelectedDate)
    TextView selectedDate;

    private final ArticleAdapter adapter = new ArticleAdapter();

    private SearchContract.Presenter presenter;

    private static final String KEY_SEARCH = "SEARCH";
    private static final String KEY_DATE = "DATE";

    public static Intent getLaunchIntent(Context from, String searchedWord, String date) {
        Intent intent = new Intent(from, ArticlesSearchActivity.class);
        intent.putExtra(KEY_SEARCH, searchedWord);
        intent.putExtra(KEY_DATE, date);
        return intent;
    }

    public static Intent getLaunchIntent(Context from, String searchWord) {
        Intent intent = new Intent(from, ArticlesSearchActivity.class);
        intent.putExtra(KEY_SEARCH, searchWord);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_articles_based_on_search_and_date);

        ButterKnife.bind(this);
        injectDependencies();

        receiveSearchDetails();
        setAdapter();
    }

    private void injectDependencies() {
        ArticlesInteractorInterface articlesInteractor = new ArticlesInteractorImpl(ApiClient.getApi());
        presenter = new SearchPresenter(articlesInteractor);
        presenter.setView(this);
    }

    public void receiveSearchDetails() {
        Intent intent = getIntent();
        presenter.setTitle(intent.getStringExtra(KEY_SEARCH), intent.getStringExtra(KEY_DATE));
    }

    @Override
    public void showTitle(String title, String date) {
        searchedWord.setText(title);
        selectedDate.setText(date);
        presenter.getSearchedArticles(title, date);
    }

    private void setAdapter() {
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);
        adapter.setOnArticleClickListener(this);
    }

    @Override
    public void onArticleClick(String url) {
        presenter.articleDetails(url);
    }

    @Override
    public void showArticles(List<Article> articles) {
        adapter.setArticles(articles);
    }

    @Override
    public void articleFailure() {
        Toast.makeText(this, R.string.error_cant_get_articles, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void startArticleDetails(String url) {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
        startActivity(intent);
    }

    @OnClick(R.id.back)
    public void goBack() {
        finish();
    }
}

File: app/src/main/java/com/example/cobe/cryptonews/ui/articleSearch/ArticlesSearchActivity.kt
package com.example.cobe.cryptonews.ui.articleSearch

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import com.example.cobe.cryptonews.R
import com.example.cobe.cryptonews.api.ApiClient.Companion.getArticleInteractor
import com.example.cobe.cryptonews.common.extensions.getIntent
import com.example.cobe.cryptonews.common.helpers.toast
import com.example.cobe.cryptonews.listeners.OnArticleClickListener
import com.example.cobe.cryptonews.model.Article
import com.example.cobe.cryptonews.presentation.SearchInterface
import com.example.cobe.cryptonews.presentation.implementation.SearchPresenterImpl
import com.example.cobe.cryptonews.ui.articles.ArticleAdapter
import kotlinx.android.synthetic.main.activity_articles_search.*

class ArticlesSearchActivity : AppCompatActivity(), SearchInterface.View, OnArticleClickListener {

    private val adapter by lazy { ArticleAdapter(this) }
    private val presenter: SearchInterface.Presenter by lazy { SearchPresenterImpl(getArticleInteractor()) }

    companion object {
        private const val KEY_SEARCH = "SEARCH"
        private const val KEY_DATE = "DATE"

        fun getLaunchIntent(from: Context, searchedWord: String, date: String) = from.getIntent<ArticlesSearchActivity>().apply {
            putExtra(KEY_SEARCH, searchedWord)
            putExtra(KEY_DATE, date)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_articles_search)
        presenter.setView(this)

        receiveSearchDetails()
        setAdapter()

        back.setOnClickListener { goBack() }
    }

    private fun receiveSearchDetails() {
        val intent = intent
        presenter.setTitle(intent.getStringExtra(KEY_SEARCH), intent.getStringExtra(KEY_DATE))
    }

    private fun setAdapter() {
        articlesSearchList.layoutManager = LinearLayoutManager(this)
        articlesSearchList.adapter = adapter
    }

    override fun onArticleClick(url: String) = presenter.articleDetails(url)

    override fun showArticles(articles: List<Article>) = adapter.setArticles(articles)

    override fun articleFailure() = toast(getString(R.string.error_cant_get_articles))

    override fun startArticleDetails(url: String) {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        startActivity(intent)
    }

    override fun showTitle(title: String, date: String) {
        searchWord.text = title
        selectedDate.text = date
        presenter.getSearchedArticles(title, date)
    }

    private fun goBack() = finish()
}
